### PR TITLE
Check if Y version cluster upgrade is allowed

### PR DESCRIPTION
**Pull request by nilsanderselde**
### Which issue this PR addresses:

Fixes [8573199](https://msazure.visualstudio.com/AzureRedHatOpenShift/_workitems/edit/8573199)

### What this PR does / why we need it:

Checks if a new query string for the admin upgrade endpoint is set to true which allows the cluster to be upgraded to the next Y version. Will need to have corresponding Geneva Action changed.

Also fixed the error checking logic in an existing unit test and corrected or clarified some comments.

### Test plan for issue:

Updated unit tests to include this new case

### Is there any documentation that needs to be updated for this PR?

Yes, cluster upgrade SOP will need to be updated

### CODE REVIEW

File: pkg/frontend/admin_openshiftcluster_upgrade.go
Comment (nilsanderselde) [inline pkg/frontend/admin_openshiftcluster_upgrade.go:33]: Open to suggestions for the naming of the parameter, value, and variable.

**Comment by mjudeikis** [comment]
> Hmm, I'm not onboard with the spec for this - why do we want a minorOK checkbox? In what circumstances are we going to check it?

This might be a bad naming.
We want to make sure we don't upgrade 4.x version when Upgrade is executed on a cluster that has 4.x latest already.

So minor in this case is Y release.

**Comment by mjudeikis** [comment]
@jim-minter you not onboard with name or approach?

**Comment by jim-minter** [comment]
I've been persuaded ;-)

File: pkg/frontend/admin_openshiftcluster_upgrade.go
Comment (mjudeikis) [inline pkg/frontend/admin_openshiftcluster_upgrade.go:33]: How does this works if I pass `True` or `TRUE`?

File: pkg/util/version/stream_test.go
Comment (mjudeikis) [inline pkg/util/version/stream_test.go:94]: I think we need to extend this test to third version:
If Y is allowed, I want to be upgraded to Y+1 and make sure I will not be upgraded to Y+2 if it is available. 
I think logic is right, but tests just to be sure. For this we might need to override 
So we have `Streams = []Stream{stream43, stream44, stream45}` 

As once 4.5 lands we will have more than 2 streams in the list.

**Review by mjudeikis — CHANGES_REQUESTED** [review]
logic looks good, just one nit for extending tests

File: pkg/frontend/admin_openshiftcluster_upgrade.go
Comment (nilsanderselde) [reply pkg/frontend/admin_openshiftcluster_upgrade.go:33]: On the Geneva side, it'll always pass `true`, but I can make this ignore case here

Comment (mjudeikis) [reply pkg/frontend/admin_openshiftcluster_upgrade.go:33]: how does it is configured on geneva side? checkbox? if so should be fine

Comment (nilsanderselde) [reply pkg/frontend/admin_openshiftcluster_upgrade.go:33]: It will be a ☑, assuming everything works as intended.

Comment (mjudeikis) [reply pkg/frontend/admin_openshiftcluster_upgrade.go:33]: Did you tested this using geneva action code and saw?

Comment (nilsanderselde) [reply pkg/frontend/admin_openshiftcluster_upgrade.go:33]: I will do that today

**Comment by mjudeikis** [comment]
@nilsanderselde gonna wait for your confirmation to merge

File: pkg/frontend/admin_openshiftcluster_upgrade.go
Comment (nilsanderselde) [reply pkg/frontend/admin_openshiftcluster_upgrade.go:33]: @mjudeikis I've verified that this boolean gets set correctly by running the GA and RP locally.

File: pkg/util/version/stream_test.go
Comment (nilsanderselde) [reply pkg/util/version/stream_test.go:106]: @mjudeikis Added

**Comment by nilsanderselde** [comment]
> @nilsanderselde gonna wait for your confirmation to merge

@mjudeikis This should be good to merge once CI and E2E are green. This new query parameter can't `== "true"` until the GA side is released, so Y upgrades will be effectively disabled until then (using this particular GA). I think that's okay, though. The GA side is ready to be reviewed.

File: pkg/util/version/stream.go
Comment (jim-minter) [inline pkg/util/version/stream.go:28]: `upgradeCandidate.Version != nil && ` is superfluous - we have already dereferenced it in the if statement above

File: pkg/util/version/stream.go
Comment (jim-minter) [inline pkg/util/version/stream.go:39]: `if v.Eq(upgradeCandidate.Version) {` is also superfluous at this point.

File: pkg/util/version/stream.go
Comment (jim-minter) [inline pkg/util/version/stream.go:29]: s/upgrading/downgrading/ might be clearer here

File: pkg/util/version/stream.go
Comment (jim-minter) [inline pkg/util/version/stream.go:43]: the following example behaviour is currently implemented, but is it actually wanted?
rp baseline = 4.4.27
cluster = 4.4.28
upgradeY = true
"not upgrading" error returned rather than upgrading the cluster to 4.5.x

File: pkg/util/version/stream.go
Comment (jim-minter) [inline pkg/util/version/stream.go:40]: this nested loop indicates that a refactor is needed.

**Review by jim-minter — CHANGES_REQUESTED** [review]
GetUpgradeStream is pretty abominable by this point.  Refactor to something like:
```golang
// getStream receives a Version x.y.z and returns the Stream x.y.0 if it exists.
func getStream(v *Version) *Stream {
	for _, s := range Streams {
		if s.Version.V[0] == v.V[0] && s.Version.V[1] == v.V[1] {
			return s
		}
	}

	return nil
}

// GetUpgradeStream returns an upgrade Stream for a Version or nil if no upgrade
// should be performed.
func GetUpgradeStream(v *Version, upgradeY bool) *Stream {
	s := getStream(v)
	if s == nil {
		return nil
	}

	if v.Lt(s.Version) {
		return s
	}

	if upgradeY {
		return getStream(&Version{V: [3]uint32{v.V[0], v.V[1] + 1}})
	}

	return nil
}
```

(in const.go, make InstallStream a *Stream, and Streams a []*Stream; in upgrade.go check `if err != nil || stream == nil { return err }` after calling GetUpgradeStream).

File: pkg/util/version/stream.go
Comment (nilsanderselde) [reply pkg/util/version/stream.go:43]: @jim-minter I don't think so -- I will check if `upgradeY` is false before returning the error above.

File: pkg/util/version/stream.go
Comment (nilsanderselde) [reply pkg/util/version/stream.go]: @jim-minter I've changed it to your suggested code

File: pkg/util/version/stream_test.go
Comment (jim-minter) [inline pkg/util/version/stream_test.go:57]: style: in general, do `foo := &Foo{}` and define foo as a pointer type at the point of definition, then you don't spread `&`s all over the rest of the codebase

File: pkg/util/version/stream_test.go
Comment (jim-minter) [inline pkg/util/version/stream_test.go:55]: style: in general, omit zero value initializers

File: pkg/util/version/stream_test.go
Comment (jim-minter) [inline pkg/util/version/stream_test.go:110]: Don't do this - messing with a global which is intended to be constant is only going to confuse someone badly later.

File: pkg/util/version/stream_test.go
Comment (jim-minter) [inline pkg/util/version/stream_test.go:120]: overkill: reflect.DeepEqual is fine with nil input; cmp.Diff will print your diff.

**Review by jim-minter — CHANGES_REQUESTED** [review]

**Review by jim-minter — APPROVED** [review]

**Pull request merged**
